Add TF lite label_image example.

thomasknauth opened the pull request:
Showcasing how to run some Tensorflow (image labeling) on Graphene.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/oscarlab/graphene-tests/22)
<!-- Reviewable:end -->

Commit:
Add TensorFlow Lite label_image example

donporter reviewed (dismissed):
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab)



<!-- Sent from Reviewable.io -->

dimakuv reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: all files reviewed, 17 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on @thomasknauth)

___
*[label_image_lite/label_image.make, line 28 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlidsV07UFBzlUuzC3Y:-LlidsV07UFBzlUuzC3Z:bl4ixyi) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.make#L28)):*
> ```Makefile
> tensorflow/configure:
> 	git clone https://github.com/tensorflow/tensorflow
> 	cd tensorflow && git checkout tags/v1.9.0
> ```

Can you specify that this is for TensorFlow version 1.9.0 in README?
___
*[label_image_lite/label_image.make, line 48 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Llie4B18FhpG7bqYyDZ:-Llie4B18FhpG7bqYyD_:b-k9puc7) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.make#L48)):*
> ```Makefile
> 
> .PHONY=clean-tmp
> clean-tmp:
> ```

Why not add to `clean` target?
___
*[label_image_lite/label_image.make, line 52 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Llie2eaE0duCu8wmHtL:-Llie2eaE0duCu8wmHtM:bhqqxla) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.make#L52)):*
> ```Makefile
> 
> .PHONY=mrproper
> mrproper: clean-tmp
> ```

Other examples use `distclean`, not `mrproper`
___
*[label_image_lite/label_image.manifest.template, line 5 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlieHKX0nlRCYkhSojr:-LlieHKX0nlRCYkhSojs:blduv4y) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L5)):*
> ```
> loader.preload = file:$(SHIMPATH)
> loader.exec = file:label_image
> loader.execname = label_image
> ```

Is `execname` needed here? If yes, I'd like a comment why
___
*[label_image_lite/label_image.manifest.template, line 6 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlifOhZ8HxF6iLvjM4t:-LlifOhZ8HxF6iLvjM4u:b81a56t) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L6)):*
> ```
> loader.exec = file:label_image
> loader.execname = label_image
> loader.env.LD_LIBRARY_PATH = /lib:/lib/x86_64-linux-gnu:/usr/lib:/usr/lib/x86_64-linux-gnu:.
> ```

Looks like you only need `/lib:.`
___
*[label_image_lite/label_image.manifest.template, line 8 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LliePRe-sFLL72cAkVP:-LliePRe-sFLL72cAkVQ:b-oq3pkj) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L8)):*
> ```
> loader.env.LD_LIBRARY_PATH = /lib:/lib/x86_64-linux-gnu:/usr/lib:/usr/lib/x86_64-linux-gnu:.
> loader.env.PATH = /bin:/usr/bin
> loader.env.USERNAME = root
> ```

This envvar needs explanation
___
*[label_image_lite/label_image.manifest.template, line 9 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlieT3S-7JqNxIKdcpV:-LlieT3S-7JqNxIKdcpW:b-oq3pkj) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L9)):*
> ```
> loader.env.PATH = /bin:/usr/bin
> loader.env.USERNAME = root
> loader.env.HOME = /
> ```

This envvar needs explanation
___
*[label_image_lite/label_image.manifest.template, line 18 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlieaN28lVLvOgcGowS:-LlieaN28lVLvOgcGowT:bucfb4s) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L18)):*
> ```
> fs.mount.lib2.type = chroot
> fs.mount.lib2.path = /lib/x86_64-linux-gnu
> fs.mount.lib2.uri = file:/lib/x86_64-linux-gnu
> ```

I don't see that you use any libs from here. Is this entry needed?
___
*[label_image_lite/label_image.manifest.template, line 22 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Lliee-YCQkE9zgga0Oj:-Lliee-YCQkE9zgga0Ok:b-eglry2) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L22)):*
> ```
> fs.mount.bin.type = chroot
> fs.mount.bin.path = /bin
> fs.mount.bin.uri = file:/bin
> ```

I don't see that you use any libs/binaries from here. Is this entry needed?
___
*[label_image_lite/label_image.manifest.template, line 26 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LliegXS5AnSzLYSQv4s:-LliegXS5AnSzLYSQv4t:b-eglry2) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L26)):*
> ```
> fs.mount.usr.type = chroot
> fs.mount.usr.path = /usr
> fs.mount.usr.uri = file:/usr
> ```

I don't see that you use any libs/binaries from here. Is this entry needed?
___
*[label_image_lite/label_image.manifest.template, line 30 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LliejPtDccG3uOdQHzu:-LliejPtDccG3uOdQHzv:b-eglry2) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L30)):*
> ```
> fs.mount.home.type = chroot
> fs.mount.home.path = /home
> fs.mount.home.uri = file:/home
> ```

I don't see that you use any libs/binaries from here. Is this entry needed?
___
*[label_image_lite/label_image.manifest.template, line 31 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Lliemc5CFpb_03bH1oJ:-Lliemc5CFpb_03bH1oK:bvpulhm) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L31)):*
> ```
> fs.mount.home.path = /home
> fs.mount.home.uri = file:/home
> sys.brk.size = 256K
> ```

Is this important? Why? Add a comment
___
*[label_image_lite/label_image.manifest.template, line 33 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LliepFS0uumI4wR5mGo:-LliepFS0uumI4wR5mGp:bnqq6jq) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L33)):*
> ```
> sys.brk.size = 256K
> 
> sgx.enclave_size = 1024M
> ```

Please a comment why this number (`# 1GB sounds like enough` is also Ok)
___
*[label_image_lite/label_image.manifest.template, line 34 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Llif-O00XjaISDiKJ1T:-Llif-O00XjaISDiKJ1U:b-bxp9t8) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L34)):*
> ```
> 
> sgx.enclave_size = 1024M
> sgx.thread_num = 16
> ```

Please a comment why this number. Why not less (if it is single-threaded)? Why not more (if it spans all availables cores, then what happens on 64-core machine)? If it depends on `-t thread-count` argument of the executable, please explain this in the comment.
___
*[label_image_lite/Makefile, line 39 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlidR_14hoePL3zC8xs:-LlidR_14hoePL3zC8xt:b-k9puc7) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/Makefile#L39)):*
> ```Makefile
> 	./pal_loader ./label_image.manifest.sgx -m inception_v3.tflite -i image.bmp -t 1
> 
> clean-tmp:
> ```

Why not add to `clean` target?
___
*[label_image_lite/Makefile, line 43 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlidM4B4KSQxXL8HImF:-LlidM4C0VHC2acr15Cr:bhqqxla) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/Makefile#L43)):*
> ```Makefile
> 
> .PHONY=mrproper
> mrproper: clean
> ```

Other examples use `distclean`, not `mrproper`
___
*[label_image_lite/README.md, line 1 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlifbhHD_09Sucm7DVF:-LlifbhHD_09Sucm7DVG:bmks60v) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/README.md#L1)):*
> ```Markdown
> This example demonstrates how to run Tensorflow Lite's label_image example on Graphene.
> ```

Please a bit more info:
1. Will it work on Ubuntu 18.04?
2. Multi-threaded support?
3. Versions tested and if newer/older version were tested and failed


<!-- Sent from Reviewable.io -->

thomasknauth commented:
> > _[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)_ status: all files reviewed, 17 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on @thomasknauth)
> 
> _[label_image_lite/label_image.make, line 28 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlidsV07UFBzlUuzC3Y:-LlidsV07UFBzlUuzC3Z:bl4ixyi) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.make#L28)):_
> 
> > ```makefile
> > tensorflow/configure:
> > 	git clone https://github.com/tensorflow/tensorflow
> > 	cd tensorflow && git checkout tags/v1.9.0
> > ```
> 
> Can you specify that this is for TensorFlow version 1.9.0 in README?

Done.

> _[label_image_lite/label_image.make, line 48 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Llie4B18FhpG7bqYyDZ:-Llie4B18FhpG7bqYyD_:b-k9puc7) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.make#L48)):_
> 
> > ```makefile
> > .PHONY=clean-tmp
> > clean-tmp:
> > ```
> 
> Why not add to `clean` target?

Changed.
 
> _[label_image_lite/label_image.make, line 52 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Llie2eaE0duCu8wmHtL:-Llie2eaE0duCu8wmHtM:bhqqxla) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.make#L52)):_
> 
> > ```makefile
> > .PHONY=mrproper
> > mrproper: clean-tmp
> > ```
> 
> Other examples use `distclean`, not `mrproper`

Changed.

> _[label_image_lite/label_image.manifest.template, line 5 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlieHKX0nlRCYkhSojr:-LlieHKX0nlRCYkhSojs:blduv4y) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L5)):_
> 
> > ```
> > loader.preload = file:$(SHIMPATH)
> > loader.exec = file:label_image
> > loader.execname = label_image
> > ```
> 
> Is `execname` needed here? If yes, I'd like a comment why

Removed.
 
> _[label_image_lite/label_image.manifest.template, line 6 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlifOhZ8HxF6iLvjM4t:-LlifOhZ8HxF6iLvjM4u:b81a56t) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L6)):_
> 
> > ```
> > loader.exec = file:label_image
> > loader.execname = label_image
> > loader.env.LD_LIBRARY_PATH = /lib:/lib/x86_64-linux-gnu:/usr/lib:/usr/lib/x86_64-linux-gnu:.
> > ```
> 
> Looks like you only need `/lib:.`

Changed.

> _[label_image_lite/label_image.manifest.template, line 8 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LliePRe-sFLL72cAkVP:-LliePRe-sFLL72cAkVQ:b-oq3pkj) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L8)):_
> 
> > ```
> > loader.env.LD_LIBRARY_PATH = /lib:/lib/x86_64-linux-gnu:/usr/lib:/usr/lib/x86_64-linux-gnu:.
> > loader.env.PATH = /bin:/usr/bin
> > loader.env.USERNAME = root
> > ```
> 
> This envvar needs explanation

Removed.
 
> _[label_image_lite/label_image.manifest.template, line 9 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlieT3S-7JqNxIKdcpV:-LlieT3S-7JqNxIKdcpW:b-oq3pkj) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L9)):_
> 
> > ```
> > loader.env.PATH = /bin:/usr/bin
> > loader.env.USERNAME = root
> > loader.env.HOME = /
> > ```
> 
> This envvar needs explanation

Removed.

> _[label_image_lite/label_image.manifest.template, line 18 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlieaN28lVLvOgcGowS:-LlieaN28lVLvOgcGowT:bucfb4s) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L18)):_
> 
> > ```
> > fs.mount.lib2.type = chroot
> > fs.mount.lib2.path = /lib/x86_64-linux-gnu
> > fs.mount.lib2.uri = file:/lib/x86_64-linux-gnu
> > ```
> 
> I don't see that you use any libs from here. Is this entry needed?

No. Removed.
 
> _[label_image_lite/label_image.manifest.template, line 22 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Lliee-YCQkE9zgga0Oj:-Lliee-YCQkE9zgga0Ok:b-eglry2) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L22)):_
> 
> > ```
> > fs.mount.bin.type = chroot
> > fs.mount.bin.path = /bin
> > fs.mount.bin.uri = file:/bin
> > ```
> 
> I don't see that you use any libs/binaries from here. Is this entry needed?

No. Removed.
 
> _[label_image_lite/label_image.manifest.template, line 26 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LliegXS5AnSzLYSQv4s:-LliegXS5AnSzLYSQv4t:b-eglry2) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L26)):_
> 
> > ```
> > fs.mount.usr.type = chroot
> > fs.mount.usr.path = /usr
> > fs.mount.usr.uri = file:/usr
> > ```
> 
> I don't see that you use any libs/binaries from here. Is this entry needed?

No. Removed.
 
> _[label_image_lite/label_image.manifest.template, line 30 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LliejPtDccG3uOdQHzu:-LliejPtDccG3uOdQHzv:b-eglry2) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L30)):_
> 
> > ```
> > fs.mount.home.type = chroot
> > fs.mount.home.path = /home
> > fs.mount.home.uri = file:/home
> > ```
> 
> I don't see that you use any libs/binaries from here. Is this entry needed?

No. Removed.
 
> _[label_image_lite/label_image.manifest.template, line 31 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Lliemc5CFpb_03bH1oJ:-Lliemc5CFpb_03bH1oK:bvpulhm) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L31)):_
> 
> > ```
> > fs.mount.home.path = /home
> > fs.mount.home.uri = file:/home
> > sys.brk.size = 256K
> > ```
> 
> Is this important? Why? Add a comment

Dunno. Removed.
 
> _[label_image_lite/label_image.manifest.template, line 33 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LliepFS0uumI4wR5mGo:-LliepFS0uumI4wR5mGp:bnqq6jq) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L33)):_
> 
> > ```
> > sys.brk.size = 256K
> > 
> > sgx.enclave_size = 1024M
> > ```
> 
> Please a comment why this number (`# 1GB sounds like enough` is also Ok)
> 
> _[label_image_lite/label_image.manifest.template, line 34 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Llif-O00XjaISDiKJ1T:-Llif-O00XjaISDiKJ1U:b-bxp9t8) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/label_image.manifest.template#L34)):_
> 
> > ```
> > 
> > sgx.enclave_size = 1024M
> > sgx.thread_num = 16
> > ```
> 
> Please a comment why this number. Why not less (if it is single-threaded)? Why not more (if it spans all availables cores, then what happens on 64-core machine)? If it depends on `-t thread-count` argument of the executable, please explain this in the comment.
> 
> _[label_image_lite/Makefile, line 39 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlidR_14hoePL3zC8xs:-LlidR_14hoePL3zC8xt:b-k9puc7) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/Makefile#L39)):_
> 
> > ```makefile
> > 	./pal_loader ./label_image.manifest.sgx -m inception_v3.tflite -i image.bmp -t 1
> > 
> > clean-tmp:
> > ```
> 
> Why not add to `clean` target?

clean_tmp and clean_extra += clean_tmp simply follow the example set by the other Makefiles. 

> _[label_image_lite/Makefile, line 43 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlidM4B4KSQxXL8HImF:-LlidM4C0VHC2acr15Cr:bhqqxla) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/Makefile#L43)):_
> 
> > ```makefile
> > .PHONY=mrproper
> > mrproper: clean
> > ```
> 
> Other examples use `distclean`, not `mrproper`

Changed.
 
> _[label_image_lite/README.md, line 1 at r1](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LlifbhHD_09Sucm7DVF:-LlifbhHD_09Sucm7DVG:bmks60v) ([raw file](https://github.com/oscarlab/graphene-tests/blob/5bfb967dd3247332db7ecacf654986487a856b6c/label_image_lite/README.md#L1)):_
> 
> > ```gfm
> > This example demonstrates how to run Tensorflow Lite's label_image example on Graphene.
> > ```
> 
> Please a bit more info:
> 
> 1. Will it work on Ubuntu 18.04?
> 2. Multi-threaded support?
> 3. Versions tested and if newer/older version were tested and failed

Done.

donporter reviewed (dismissed):
> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: all files reviewed, 17 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: OSCAR Lab) (waiting on @thomasknauth)



<!-- Sent from Reviewable.io -->

dimakuv reviewed (dismissed):
> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: :shipit: complete! all files reviewed, all discussions resolved

___
*[label_image_lite/README.md, line 5 at r2](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LmLO4hb2KfLakYvSboV:-LmLO4hb2KfLakYvSboW:brmrf1u) ([raw file](https://github.com/oscarlab/graphene-tests/blob/fc507ea882d2ff86e627b7f1c5551ba28b68893b/label_image_lite/README.md#L5)):*
> ```Markdown
> Known limitations:
> 
> - Tested on Ubuntu 16.04 with Graphene's chiache/glibc-2.23 branch. Building Tensorflow on Ubuntu 16.04 creates dependencies on this newer glibc. Ubuntu 18.04 should work, but have not tested.
> ```

BTW: We should change this sentence ("chiache/glibc... branch") after we merge PR 809 and test that this build works with the newest Graphene master.


<!-- Sent from Reviewable.io -->

dimakuv reviewed (dismissed):
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel)



<!-- Sent from Reviewable.io -->

donporter reviewed (dismissed):
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

mkow reviewed (changes requested):
> Reviewed 3 of 4 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: all files reviewed, 5 unresolved discussions (waiting on @thomasknauth)

___
*[label_image_lite/label_image.make, line 5 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LmfkN0p56i9RYAZWw2b:-LmfkN0q1CLgInt3LvVf:b-y2vimj) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/label_image.make#L5)):*
> ```Makefile
> BAZEL_BIN=$(HOME)/bin/bazel
> 
> .PHONY=default
> ```

Shouldn't this be `.PHONY: default`? Same for all other places in this file.
___
*[label_image_lite/label_image.manifest.template, line 18 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Lmfl1TFE8iN1dvqQEMG:-Lmfl1TG5oROz-8oqNTK:b-pmwkey) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/label_image.manifest.template#L18)):*
> ```
> # 2 threads are sufficient when telling label_image to only use a single thread (-t 1 command line
> # parameter; see Makefile's check target).
> sgx.thread_num = 2
> ```

I don't think this is enough, Graphene uses up to 2 helper threads which counts towards this limit (though it's not yet documented anywhere).
___
*[label_image_lite/Makefile, line 37 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LmflPmBErZ_rd2tj-CF:-LmflPmBErZ_rd2tj-CG:b-7ropln) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/Makefile#L37)):*
> ```Makefile
> 	cp /usr/lib/x86_64-linux-gnu/libstdc++.so.6 $@
> 
> .PHONY=check
> ```

ditto
___
*[label_image_lite/Makefile, line 41 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Lmfl_Um-zWbIyppvACq:-Lmfl_Um-zWbIyppvACr:b-j8q1fx) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/Makefile#L41)):*
> ```Makefile
> 	./pal_loader ./label_image.manifest.sgx -m inception_v3.tflite -i image.bmp -t 1
> 
> clean-tmp:
> ```

missing .PHONY
___
*[label_image_lite/Makefile, line 44 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LmflbBF94e68trH1Qm_:-LmflbBF94e68trH1Qma:b-7ropln) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/Makefile#L44)):*
> ```Makefile
> 	$(MAKE) -f label_image.make clean
> 
> .PHONY=distclean
> ```

ditto


<!-- Sent from Reviewable.io -->

thomasknauth reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: 2 of 4 files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv, @donporter, and @mkow)

___
*[label_image_lite/label_image.make, line 5 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LmfkN0p56i9RYAZWw2b:-LmlKTsiBgQu4bc2CxUG:b-896fix) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/label_image.make#L5)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Shouldn't this be `.PHONY: default`? Same for all other places in this file.
</blockquote></details>

Done.
___
*[label_image_lite/label_image.manifest.template, line 18 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Lmfl1TFE8iN1dvqQEMG:-LmlKYGe48dPs6jku0WR:b4zhkwb) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/label_image.manifest.template#L18)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

I don't think this is enough, Graphene uses up to 2 helper threads which counts towards this limit (though it's not yet documented anywhere).
</blockquote></details>

Dunno, it ran without errors for me.
___
*[label_image_lite/Makefile, line 37 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LmflPmBErZ_rd2tj-CF:-LmlKcMcDpV-3KxG7TSt:b-896fix) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/Makefile#L37)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[label_image_lite/Makefile, line 41 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Lmfl_Um-zWbIyppvACq:-LmlKpFK7ULuNa7G1GOb:b-896fix) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/Makefile#L41)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

missing .PHONY
</blockquote></details>

Done.
___
*[label_image_lite/Makefile, line 44 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-LmflbBF94e68trH1Qm_:-LmlKrGZCzTwA4vkjW7k:b-896fix) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/Makefile#L44)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mkow reviewed (changes requested):
> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @thomasknauth)

___
*[label_image_lite/label_image.manifest.template, line 18 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Lmfl1TFE8iN1dvqQEMG:-Lmob3_B68pHRQTtd9ta:bbuegcj) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/label_image.manifest.template#L18)):*
<details><summary><i>Previously, thomasknauth (Thomas Knauth) wrote…</i></summary><blockquote>

Dunno, it ran without errors for me.
</blockquote></details>

Those additional threads are spawned on-demand, so it may sometimes work with smaller limits. I think this should always be max_app_threads + 2 to be safe (and stable).


<!-- Sent from Reviewable.io -->

mkow reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @thomasknauth)

___
*[label_image_lite/label_image.manifest.template, line 18 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Lmfl1TFE8iN1dvqQEMG:-LmpTPro2_yc-nqTZkYM:b-knjaxg) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/label_image.manifest.template#L18)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Those additional threads are spawned on-demand, so it may sometimes work with smaller limits. I think this should always be max_app_threads + 2 to be safe (and stable).
</blockquote></details>

For a reference, please take a look at this: https://github.com/oscarlab/graphene-tests/blob/35d76947e245d1a3b08c5c29eea51c0dd6497f1f/memcached/memcached.manifest.template#L126-L130


<!-- Sent from Reviewable.io -->

thomasknauth reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow)

___
*[label_image_lite/label_image.manifest.template, line 18 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Lmfl1TFE8iN1dvqQEMG:-LmpU5vc1eJdG9iTzbjI:b68fnfn) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/label_image.manifest.template#L18)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

For a reference, please take a look at this: https://github.com/oscarlab/graphene-tests/blob/35d76947e245d1a3b08c5c29eea51c0dd6497f1f/memcached/memcached.manifest.template#L126-L130
</blockquote></details>

I see your point, but do we want to put this exclaimer in every Makefile? Seems like this is a Graphene internal thing. Maybe that number should really only refer to the threads available to the application? I understand why it's needed in SGXv1, but as a user of Graphene, I don't really want to think about how many internal threads Graphene might spawn and I have to account for.


<!-- Sent from Reviewable.io -->

thomasknauth reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: 3 of 4 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv, @donporter, and @mkow)

___
*[label_image_lite/label_image.manifest.template, line 18 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Lmfl1TFE8iN1dvqQEMG:-LmpWCAW8UFe-MYpAnRk:b-896fix) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/label_image.manifest.template#L18)):*
<details><summary><i>Previously, thomasknauth (Thomas Knauth) wrote…</i></summary><blockquote>

I see your point, but do we want to put this exclaimer in every Makefile? Seems like this is a Graphene internal thing. Maybe that number should really only refer to the threads available to the application? I understand why it's needed in SGXv1, but as a user of Graphene, I don't really want to think about how many internal threads Graphene might spawn and I have to account for.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mkow reviewed (dismissed):
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL)

___
*[label_image_lite/label_image.manifest.template, line 18 at r3](https://reviewable.io/reviews/oscarlab/graphene-tests/22#-Lmfl1TFE8iN1dvqQEMG:-LmpgSaO0bxEzoo-JNAx:blrfbho) ([raw file](https://github.com/oscarlab/graphene-tests/blob/25fd49c0cdaf987ed1deb970b7e80abb54e4adc6/label_image_lite/label_image.manifest.template#L18)):*
> but do we want to put this exclaimer in every Makefile?

I haven't said that ;) Just that the current calculations included here weren't correct ;)

> Seems like this is a Graphene internal thing. Maybe that number should really only refer to the threads available to the application? I understand why it's needed in SGXv1, but as a user of Graphene, I don't really want to think about how many internal threads Graphene might spawn and I have to account for.

I totally agree, this is something we need to fix. But that's something for another PR which would actually change the meaning of this parameter.


<!-- Sent from Reviewable.io -->

mkow reviewed (approved):
> Reviewed 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL)



<!-- Sent from Reviewable.io -->

donporter reviewed (approved):
> Reviewed 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

yamahata reviewed (approved):
> Reviewed 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-tests/22)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->
